### ISSUE TITLE
Better precision Trigonometry Functions

### ISSUE THREAD

**Issue by AncientHodler** [shopspring/decimal#203]
Ive seen you have implemented sin, cos etc trigonometri functions.
Their implementation is pretty complex.

However i have written for mysefl a shorter better version, based on taylor series expansions.
It has the same precision as its python equivalent, and this has the same precision as the result given by wolphram alpha.

I recommend you use them as well.
Its really easy to implement, you just need to "rethink" python code to work with the decimal syntax you guys have been creating.

The python functions are given in their decimal page. Here:
https://docs.python.org/3/library/decimal.html

Just scroll a bit down and youll see them.

**Comment by mwoss** [comment]
Does the current implementation of those trigonometric functions not work correctly? Or you meant the current implementation is just too complex and it can be easily refactored?
If the second option, I would add this suggestion to our backlog and work on this after v.1.3.0 release.

**Comment by AncientHodler** [comment]
They work, but their result is a bit off of the python code/go ported python code/wolphram alpha (all 3 of which give the exact decimal value up to 100+ and even 200+ decimal)
You implementation is compared to this precise up to the 15th decimal (up to the 15 decimal shopspring implementation is similar to the above 3 version), after which the decimals are completly differenty, so i would assume, they are incorectly calculated.
As i needed 250 precision for the function, i wrote the python equivalent in golang using the shopspring decimal computing syntax. Which solved my problem.
For now i wrote the sinus function, but ill be writing all the trigonometry functions, pi, radians, and even arccos variants...

I havent managed to follow up what logic you guys used for your functions, but i believe these other versions are easier to write and more precise.

Im building a blockchain, and instead of the standard blockchain implementations where numbers are store as int/uint types, with their inheret limitations (see ethereum overflow bug) i intend to do all the math in decimal. It may mean a bit more storage, and slowness is involved, but at least it will be precise like humans expect numbers to be precise when they add them up, and i wont have a maximum type limit above which no value can be saved.

Ill also be needing logarithms, so ill need to implement those as well.
I initially wrote some math code in python as its easier to create there, to prototype it, but python isnt suited as language for a blockchain, hence golang. Golang doesnt have native decimal type (cant imagine why it doesnt...) , hence your program.

Im "taking" only the functions i need, and add/modify those functions that i believe i have a better solution for. Ill also need to add a few more that are missing.

**Comment by mwoss** [comment]
Got it, I think I understand your use case.
I will try to move Pow implementation further next week, but I can't promise anything as I'm working on this library after work and usually I feel mentally exhausted after 8h of working in the current job :||

Btw about your use case. It seems you are doing computational intensive operations in your blockchain implementation. As I understand your approach and the problem you want to solve, have you ever consider even more performant languages such as Rust or bare C? I think

**Comment by mwoss** [comment]
Also if you wonder why Golang does not have a built-in decimal type I'm linking below few issues where community was discussing that with Golang authors:
- https://github.com/golang/go/issues/12127
- https://github.com/golang/go/issues/12332

**Comment by AncientHodler** [comment]
I dont know about that, but whatever trickery this module does with the golang built in int and float types, sure looks like working decimal numbers to me.
Saying we wont build something because its impossible its just plain stupid. this module shows that decimal math can exist in golang. And ive written a sqrt function as well using the babilonian method. it computes perfectly sqrt from 2 with 250 decimal precision in less than 10 loops. what more can i ask.
I know, a function that computes logarithms, im gonna need to write that as well..

Im not sure how big memory wise, such data types are.
Im e create  a decimal from string with 1000numbers before the decimal point and 1000 numbers after the decimal point, it shows the same size as a number like 2131.2131, ....

**Comment by AncientHodler** [comment]
I got it now how the python trigonometric loop stop. they stop based on precision. i can easily implement that using truncate, as your library cant set a certain context precision, upon which math operates, or am i wrong ?

**Comment by AncientHodler** [comment]
> 
> 
> Got it, I think I understand your use case.
> I will try to move Pow implementation further next week, but I can't promise anything as I'm working on this library after work and usually I feel mentally exhausted after 8h of working in the current job :||
> 
> Btw about your use case. It seems you are doing computational intensive operations in your blockchain implementation. As I understand your approach and the problem you want to solve, have you ever consider even more performant languages such as Rust or bare C? I think
No, im sticking to golang. The computation intensive part spits some high precision decimal numbers, which will then be used as seeds as inputs into another code that is not that computational intesive. You library suits me just fine. Dont need to rush with the POW function... i wrote me a babilonian method for sqrt that works perfectly. it basically does pow(0.5), its what i needed. If i need more complicated stuff il just write them myself. I needed to solve an equation in one of my other programs that involved using a lamber w function. but using the function from python library didnt yield me the results i needed, so i ended up writing an algo that computed me the solution using a logic a devised myself, which worked as intended, solving the equation to the 24th decimal, which is what i needed. Im looking forward in porting that to golang using your decimal module, however, ill need logarithm function for that...

**Comment by mwoss** [comment]
> Pythons implementation spwed some weird dhiet out like infinity + NotANumber * i, wierdest shiet i ever witnessed to be hones
Haha, I  laughed so hard reading this xD

Btw. About that logarithm function. I think I have a draft of the `Ln` function somewhere on the local branch on the other PC. Do you need a natural logarithm for your calculations or a logarithm function where you could pass any logarithmic base?

**Comment by AncientHodler** [comment]
You might want to check this:
https://arxiv.org/pdf/1410.7176 titled
Efficient implementation of elementary functions in the medium-precision range
Elementary functions being
exp, sin, cos, log, atan

"We describe a new implementation of the elementary transcendental functions exp,
sin, cos, log and atan for variable precision up to approximately 4096 bits. Compared
to the MPFR library, we achieve a maximum speedup ranging from a factor 3 for cos to
30 for atan. Our implementation uses table-based argument reduction together with
rectangular splitting to evaluate Taylor series. We collect denominators to reduce the
number of divisions in the Taylor series, and avoid overhead by doing all multiprecision
arithmetic using the mpn layer of the GMP library. Our implementation provides
rigorous error bounds."
Im reading the pdf, and looking if i can implement this myself using your library math syntax.

If we have the elementary functions implemented like this, it would be a major, win, as wed have logarithm, power function, and all trigonometric functions.

**Comment by AncientHodler** [comment]
> 
> 
> > Pythons implementation spwed some weird dhiet out like infinity + NotANumber * i, wierdest shiet i ever witnessed to be hones
> 
> Haha, I laughed so hard reading this xD
> 
> Btw. About that logarithm function. I think I have a draft of the `Ln` function somewhere on the local branch on the other PC. Do you need a natural logarithm for your calculations or a logarithm function where you could pass any logarithmic base?

i need to do do log in base x from y, which basically boils down to knowing log in base e from x.

**Comment by AncientHodler** [comment]
The source code for these implementation as describe in the paper is here :
https://github.com/fredrik-johansson/arb
But its foreign c++ language ....

**Comment by AncientHodler** [comment]
> 
> 
> The source code for these implementation as describe in the paper is here :
> https://github.com/fredrik-johansson/arb
> But its foreign c++ language ....
But they did code some messed up shiet, check it here
https://arblib.org/examples.html

**Comment by AncientHodler** [comment]
> > Pythons implementation spwed some weird dhiet out like infinity + NotANumber * i, wierdest shiet i ever witnessed to be hones
> 
> Haha, I laughed so hard reading this xD
> 
> Btw. About that logarithm function. I think I have a draft of the `Ln` function somewhere on the local branch on the other PC. Do you need a natural logarithm for your calculations or a logarithm function where you could pass any logarithmic base?

Yeah, id ver much like to see your implementation. I wrote code that computes the natural logaritm and it works for the small number with reasonable precision. But if i kick it into overdrive, for what i need, for instance, to compute the logaritm in base 7777777777777 from 98375453234123.235645874563254254856325 it would probably take a weeks if not months. i dont know how the hell does python do it, cause it works instant.....

https://play.golang.org/p/OWh05UeoBkL

**Comment by mwoss** [comment]
Python implementation of maths operations (in decimal lib) are a bit tricky sometimes and hard to understand as authors usually rely on Python language features. I've tried to base my initial Pow implementation on their implementation and I've completely failed. Too many "Python" only workarounds were used there.
